### DISCUSSION TITLE
Mikro ORM not handling many to many relation correctly

### DISCUSSION THREAD

**Discussion by Marius-Le-Houezec-Lacaze** [mikro-orm/mikro-orm#3663] [Q&A]
**Describe the bug**
Mikro orm does not handle many to many correctly, in this situation it join `cc.users` as `clientCompany` instead of `user` wich it should be  

**To Reproduce**
```ts
   let qb2 = this.buildingQueryRepository
      .createQueryBuilder('fr')
      .select('fr.id')
      .leftJoin('fr.building', 'f')
      .leftJoin('f.clientCompany', 'cc')
      .leftJoin('cc.users', 'ucc')
      .where('ucc.username="test"')
```

`ClientCompanyEntity`
```ts
  @ManyToMany({ entity: () => User, pivotEntity: () => UserHasClientCompany})
  users = new Collection<User>(this);
```

`SQL output`
```sql
select `fr`.`id` from `buildingReport` as `fr` left join `building` as `f` on `fr`.`buildingId` = `f`.`id` left join `clientCompany` as `cc` on `f`.`clientCompanyId` = `cc`.`id` left join `userHasClientCompany` as `u1` on `cc`.`id` = `u1`.`clientCompanyId` left join `clientCompany` as `ucc` on `u1`.`clientCompanyId` = `ucc`.`id` where (ucc.username="test")
```

**Expected behavior**
Mikro orm should join the many to many relation cc.users on the table users instead of the table clientCompany.:

this :
```
left join `clientCompany` as `ucc`
```
should be:
```
left join `cc.users` as `ucc`
```
or something similar

**Versions**

| Dependency | Version |
| - | - |
| node | v16.15.0 |
| mikro-orm | ^5.5.0 |

**Comment by B4nan** [comment]
Please provide complete repro, this is not enough.

**Reply by Marius-Le-Houezec-Lacaze** [reply]
Very well i'll try to provide you with one ASAP

**Comment by Marius-Le-Houezec-Lacaze** [comment]
Here i created a project to replicate the bug easily simply go to `/building` route and the error will be very apparent in the console 

[Repo to recreate the bug](https://github.com/Marius-Le-Houezec-Lacaze/Mikro-ORM-many-to-many-issue)


``` sql
Error: select `bd`.`id` from `building` as `bd` left join `clientCompany` as `cc` on `bd`.`id` = `cc`.`buildingId` left join `userHasClientCompany` as `u1` on `cc`.`id` = `u1`.`clientCompanyId` left join `clientCompany` as `ucc` on `u1`.`clientCompanyId` = `ucc`.`id` where (ucc.username="test") - Unknown column 'ucc.username' in 'where clause'
```

it tries to join clientCompany has ucc instead of user as ucc

**Comment by Marius-Le-Houezec-Lacaze** [comment]
@B4nan Sorry to bump any update on this issue ? Do you need anything else maybe ?

**Reply by B4nan** [reply]
Would be great if you can simplify it, your repo is a complex web app, not a minimal reproduction. Ideally, it should be a simple script, I don;t want to browse through many files. Check existing [regression tests](https://github.com/mikro-orm/mikro-orm/tree/master/tests/issues) for inspiration, or use this https://stackblitz.com/edit/mikro-orm-repro as a base for it.

I can take a look at the one you provided too, but I need to prioritize things I can easily reproduce, and there are too many of them right now, so it might take few more days - I've been out most of the last week, so I have quite some things to catch up with.

I see you are using many customizations, including your own naming strategy, those things should be the first that you try to remove, to ensure it's not about that. The final repro should use as many defaults as possible. During the process of simplification, you often find the cause yourself, which will let me focus on the actual fix instead of simplifying complex repros :]

**Reply by Marius-Le-Houezec-Lacaze** [reply]
Sorry i am quite busy also i'll try to simplify it i am not used to MikroORM standalone outisde of nestjs i'll see if i can provide you with a simplified repo tommorow

**Reply by Marius-Le-Houezec-Lacaze** [reply]
Well if i don't have the time i don't mind if you can look at this repo it should not be too complicated just running npm install and the migration
